src/java/voldemort/server/protocol/admin/AdminServiceRequestHandler.java

```diff
@@ -1085,14 +1093,35 @@ private String swapStore(String storeName, String directory) throws VoldemortExc
                 logger.warn("Push Version is not specified, this might create issues during rebalance/restore. Store"
                         + storeName + " Generated version " + pushVersion);
             }
+
+            FileBackedCachingStorageEngine quotaStore = (FileBackedCachingStorageEngine)
+                    storeRepository.getStorageEngine(SystemStoreConstants.SystemStoreName.voldsys$_store_quotas.toString());
+            String diskQuotaSize = quotaStore.cacheGet(storeName);
+            Long diskQuotaSizeInKB = (diskQuotaSize == null) ? null : Long.parseLong(diskQuotaSize);
+
             ReadOnlyStoreFetchOperation operation = new ReadOnlyStoreFetchOperation(requestId,
                                                                                     metadataStore,
                                                                                     store,
                                                                                     fileFetcher,
                                                                                     storeName,
                                                                                     fetchUrl,
-                                                                                    pushVersion);
-            asyncService.submitOperation(requestId, operation);
+                                                                                    pushVersion,
+                                                                                    diskQuotaSizeInKB);
+            AdminServiceRequestHandler.storeLock.lock();
+
+            try {
+                if (AdminServiceRequestHandler.fetchingStores.containsKey(storeName)) {
+                    int previousRequestID = AdminServiceRequestHandler.fetchingStores.get(storeName);
+                    if (!asyncService.isComplete(previousRequestID, false))
+                        throw new VoldemortException("The store: " + storeName + " is currently block since it is fetching data");
```

Comment:
Typo: block => blocked

Since you have the request ID that is currently in progress, I suggest including it in the exception message. That would be very useful information, I think.